src/monitor/group_state_machine.c

```diff
@@ -678,7 +682,11 @@ ProceedGroupState(AutoFailoverNode *activeNode)
 
 		/* it's safe to rejoin as a secondary */
 		AssignGoalState(activeNode, REPLICATION_STATE_SECONDARY, message);
-		AssignGoalState(primaryNode, REPLICATION_STATE_PRIMARY, message);
+
+		if (activeNode->candidatePriority > 0)
```

Comment:
I cannot follow this logic. Why do we check the  newly  transitioned secondary's candidate priority for promoting the other node? Or, maybe conversely, why would we NOT promote the primary if `activeNode->candidatePriority = 0`?

Reply:
The state PRIMARY embeds the assumption that it's possible to failover. A node with candidate Priority of zero can not be the target of a failover, so might reach SECONDARY without allowing to get out of WAIT_PRIMARY or JOIN_PRIMARY state.

At some point in the past I had a proposal to introduce a STREAMING state that would be like a SECONDARY state, but without being a target for promotion. Maybe that would be a good way to address the current situation...

Reply:
I really cannot understand this part well. I'm not saying that this is wrong or anything, but not intuitive for me and I really like to understand  better before approving.

1. This code-path is triggered only for single-standby case. Do we  ever care  about `candidatePriority ` in the single-standby case?
2.
>  A node with candidate Priority of zero can not be the target of a failover

Why do  we care about `activeNode `'s (e.g., the secondary's) priority when it is assigned to be secondary at the moment?  even if it is zero, why  don't we just allow `activeNode` to be  the `REPLICATION_STATE_SECONDARY` and primaryNode to be  in  `REPLICATION_STATE_PRIMARY`?

3. > so might reach SECONDARY without allowing to get out of WAIT_PRIMARY or JOIN_PRIMARY state.

The  `activeNode`  reaches `REPLICATION_STATE_SECONDARY ` and `primaryNode ` reaches to `REPLICATION_STATE_PRIMARY ` right here. When/why wouldn't it  allow to  get out of `WAIT_PRIMARY` or `JOIN_PRIMARY`?


4. Optionally, can you give an example flow where this change fixes the problem?

Reply:
This transition happens when a former primary node has been demoted, and is now accepted back in the group as a secondary. 

  - When the group has only two nodes, the other node must have been promoted, and is now in the state WAIT_PRIMARY.

  - When the group has more than two nodes, an election is taking place while the former primary is demoted. One of the other nodes is going to be selected and assigned PREPARE_PROMOTION, and then later may reach PRIMARY when one of the still healthy secondaries follow the new primary.

The situation we are interested in with this PR is the situation when the new primary node can be assigned PRIMARY again, meaning the group is now ready for another failover if necessary. If the first standby that follows the new primary is not a candidate for promotion, we can't move to PRIMARY yet, we need to wait until a secondary that is a candidate for promotion shows up.

Reply:
> This transition happens when a former primary node has been demoted, and is now accepted back in the group as a secondary.

Alright, now  I got it  thanks. Can you maybe  add this to  the comment in the code?